File: agent/market_regime.py
"""Gate #1 — Market Regime. Day 3 weekly checklist + Day 2 VIX overlay."""
from __future__ import annotations
from dataclasses import dataclass, field
from . import data
from .technicals import ema


@dataclass
class Regime:
    score: int = 0
    checks: dict = field(default_factory=dict)
    mode: str = "NO_TRADE"          # AGGRESSIVE | CAUTIOUS | NO_TRADE
    size_multiplier: float = 0.0    # 1.0 / 0.5 / 0.0
    vix: float | None = None
    vix_note: str = ""
    spy_close: float | None = None


def assess() -> Regime:
    r = Regime()
    spy = data.history("SPY", period="2y")
    qqq = data.history("QQQ", period="6mo")
    vix = data.history("^VIX", period="3mo")

    if spy.empty or qqq.empty:
        r.checks["error"] = "no index data"
        return r

    # 1. SPY > EMA200
    spy_ema200 = ema(spy["Close"], 200).iloc[-1]
    r.spy_close = float(spy["Close"].iloc[-1])
    c1 = r.spy_close > spy_ema200
    r.checks["SPY > EMA200"] = c1

    # 2. QQQ new 4-week high within last 5 sessions
    q = qqq["Close"]
    rolling_high = q.rolling(20).max()
    c2 = bool((q.iloc[-5:] >= rolling_high.iloc[-5:] * 0.999).any())
    r.checks["QQQ 4-week high (last 5d)"] = c2

    # 3. VIX < 20
    r.vix = float(vix["Close"].iloc[-1]) if not vix.empty else None
    c3 = r.vix is not None and r.vix < 20
    r.checks["VIX < 20"] = c3

    # 4. SPY up on the week (last 5 sessions vs prior 5)
    w = spy["Close"].iloc[-11:]
    c4 = bool(w.iloc[-1] > w.iloc[-6]) if len(w) >= 11 else False
    r.checks["SPY up on week"] = c4

    r.score = sum(bool(x) for x in (c1, c2, c3, c4))
    if r.score == 4:
        r.mode, r.size_multiplier = "AGGRESSIVE", 1.0
    elif r.score >= 2:
        r.mode, r.size_multiplier = "CAUTIOUS", 0.5
    else:
        r.mode, r.size_multiplier = "NO_TRADE", 0.0

    if r.vix is not None:
        if r.vix > 28:
            r.vix_note = ("PANIC ZONE (VIX>28): contrarian window — scale into "
                          "fundamentally strong names in parts (15%/40%), never all at once.")
        elif r.vix < 15:
            r.vix_note = "Calm (VIX<15): keep booking profits per pie-exit plan."
        else:
            r.vix_note = "VIX 15–28: hold with the trend, monitor."
    return r

File: web/lib/marketRegime.ts
/** Port of agent/market_regime.py -- Gate #1, Day 3 weekly checklist + VIX overlay. */
import { fetchDailyBars } from "./marketData";
import { ema } from "./technicals";

export interface Regime {
  score: number;
  checks: Record<string, boolean>;
  mode: "AGGRESSIVE" | "CAUTIOUS" | "NO_TRADE";
  size_multiplier: number;
  vix: number | null;
  vix_note: string;
}

export async function assessRegime(): Promise<Regime> {
  const [spy, rsp, vix] = await Promise.all([
    fetchDailyBars("SPY", "2y"),
    fetchDailyBars("RSP", "6mo"),
    fetchDailyBars("^VIX", "3mo"),
  ]);

  const checks: Record<string, boolean> = {};
  if (spy.length === 0 || rsp.length === 0) {
    return { score: 0, checks: { error: false }, mode: "NO_TRADE", size_multiplier: 0, vix: null, vix_note: "" };
  }

  const spyCloses = spy.map((b) => b.close);
  const spyEma200 = ema(spyCloses, 200);
  const c1 = spyCloses[spyCloses.length - 1] > spyEma200[spyEma200.length - 1];
  checks["SPY > EMA200"] = c1;

  // Breadth: RSP (equal-weight S&P 500) vs SPY (cap-weight) over 20 trading
  // days. If RSP keeps pace, participation is broad and healthy; if it
  // meaningfully lags, the market is being carried by a handful of
  // mega-caps -- a fragility signal the old "QQQ new 4-week high" check
  // (narrow, binary, redundant with the SPY-up-this-week check below)
  // didn't capture. 1% tolerance band absorbs day-to-day noise.
  const spyByDate = new Map(spy.map((b) => [b.date, b.close]));
  const ratioSeries = rsp.filter((b) => spyByDate.has(b.date)).map((b) => b.close / (spyByDate.get(b.date) as number));
  const c2 = ratioSeries.length > 20 && ratioSeries[ratioSeries.length - 1] >= ratioSeries[ratioSeries.length - 21] * 0.99;
  checks["Breadth: RSP keeping pace with SPY (20d)"] = c2;

  const vixClose = vix.length > 0 ? vix[vix.length - 1].close : null;
  const c3 = vixClose !== null && vixClose < 20;
  checks["VIX < 20"] = c3;

  const w = spyCloses.slice(-11);
  const c4 = w.length >= 11 && w[w.length - 1] > w[w.length - 6];
  checks["SPY up on week"] = c4;

  const score = [c1, c2, c3, c4].filter(Boolean).length;
  let mode: Regime["mode"] = "NO_TRADE";
  let sizeMultiplier = 0;
  if (score === 4) { mode = "AGGRESSIVE"; sizeMultiplier = 1.0; }
  else if (score >= 2) { mode = "CAUTIOUS"; sizeMultiplier = 0.5; }

  let vixNote = "";
  if (vixClose !== null) {
    if (vixClose > 28) vixNote = "PANIC ZONE (VIX>28): contrarian window -- scale in slowly.";
    else if (vixClose < 15) vixNote = "Calm (VIX<15): keep booking profits per pie-exit plan.";
    else vixNote = "VIX 15-28: hold with the trend, monitor.";
  }

  return { score, checks, mode, size_multiplier: sizeMultiplier, vix: vixClose, vix_note: vixNote };
}
